File: 卡码/代码随想录/1/07-回溯/Go/01-组合/LeetCode_0040.go
// 40. 组合总和 II
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	path:=make([]int,0)
	res:=make([][]int,0)
	var backtracking func(candidates []int,target,sum,index int)
	backtracking=func(candidates []int,target,sum,index int){
		if sum==target {
			tmp:=make([]int,len(path))
			copy(tmp,path)
			res=append(res,tmp)
			return
		}
		for i:=index;i<len(candidates)&&sum+candidates[i]<=target;i++ {
			if i>index&&candidates[i]==candidates[i-1] {
				continue
			}
			path=append(path,candidates[i])
			backtracking(candidates,target,sum+candidates[i],i+1)
			path=path[:len(path)-1]
		}
	}
	sort.Slice(candidates,func(i,j int) bool{
		return candidates[i]<candidates[j]
	})
	backtracking(candidates,target,0,0)
	return res
}

File: 卡码/代码随想录/1/07-回溯/Python/01-组合/LeetCode_0040.py
#40. 组合总和 II
from typing import *

class Solution:
  def __init__(self):
    self.path:List[int]=[]
    self.res:List[List[int]]=[]
  def backtracking(self,candidates:List[int],target:int,sum:int,index:int):
    if sum==target:
      self.res.append(self.path[:])
      return
    for i in range(index,len(candidates)):
      if sum+candidates[i]>target:
        break
      if i>index and candidates[i]==candidates[i-1]:
        continue
      self.path.append(candidates[i])
      self.backtracking(candidates,target,sum+candidates[i],i+1)
      self.path.pop()
  def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
    candidates.sort()
    self.backtracking(candidates,target,0,0)
    return self.res